default/AI/ShipDesignAI.py

```diff
@@ -0,0 +1,1602 @@
+                    design_id = next(iter([shipDesignID for shipDesignID in fo.getEmpire().allShipDesigns
+                                          if designname == fo.getShipDesign(shipDesignID).name(False)]), None)
+                    if design_id is not None:
+                        self.design_id_by_name[designname] = design_id
+                    else:
+                        corrupted.append(designname)
+            except AttributeError:
+                print "WARNING: ShipID cache corrupted. Could not get cached shipdesign. Repairing Cache."
+                print traceback.format_exc()  # do not print to stderr as this is an "expected" exception.
+                design_id = next(iter([shipDesignID for shipDesignID in fo.getEmpire().allShipDesigns
+                                      if designname == fo.getShipDesign(shipDesignID).name(False)]), None)
+                if design_id is not None:
+                    self.design_id_by_name[designname] = design_id
+                else:
+                    corrupted.append(designname)
+        for corrupted_entry in corrupted:
+            del self.design_id_by_name[corrupted_entry]
+
+    def _update_buildable_items_this_turn(self, verbose=False):
+        """Calculate which parts and hulls can be built on each planet this turn.
+
+        :param verbose: bool - toggles detailed debugging output.
+        """
+        # The AI currently has no way of checking building requirements of individual parts and hulls directly.
+        # It can only check if we can build a design. Therefore, we use specific testdesigns to check if we can
+        # build a hull or part.
+        # The building requirements are constant so calculate this only once at the beginning of each turn.
+        #
+        # Code structure:
+        #   1. Update hull test designs
+        #   2. Get a list of buildable ship hulls for each planet
+        #   3. Update ship part test designs
+        #   4. Cache the list of buildable ship parts for each planet
+        #
+        self.hulls_for_planets.clear()
+        self.parts_for_planets.clear()
+        get_shipdesign = fo.getShipDesign
+        get_hulltype = fo.getHullType
+        empire = fo.getEmpire()
+        empire_id = empire.empireID
+        universe = fo.getUniverse()
+        planets_with_shipyards = _get_planets_with_shipyard()
+        available_hulls = list(empire.availableShipHulls)   # copy so we can sort it locally
+        try:
+            idx = available_hulls.index(TESTDESIGN_PREFERED_HULL)
+            available_hulls[0], available_hulls[idx] = available_hulls[idx], available_hulls[0]
+        except ValueError:
+            print "WARNING: Tried to use '%s' as testhull but it is not in available_hulls." % TESTDESIGN_PREFERED_HULL,
+            print "Please update ShipDesignAI.py according to the new content."
+            traceback.print_exc()
+        testdesign_ids = [shipDesignID for shipDesignID in empire.allShipDesigns
```

Comment:
You don't need  `testdesign_ids` here, you need `testdesign_names`   this will simplify next usages.